File: 310-minimum-height-trees/minimum-height-trees.py
import numpy as np

class Solution:
    def findMinHeightTrees(self, n: int, edges: List[List[int]]) -> List[int]:
        # Special cases for edge cases in the input
        if n == 1:
            return [0]  # Only one node, it is the root and the only result
        if not edges:
            return []  # No edges provided, no trees can be formed

        # Build an adjacency list to represent the graph
        edgesdict = collections.defaultdict(list)
        for edge1, edge2 in edges:
            # Add each connection in both directions (undirected graph)
            edgesdict[edge1].append(edge2)
            edgesdict[edge2].append(edge1)

        # Initialize a list of leaf nodes (nodes with only one neighbor)
        leaves = [key for key, value in edgesdict.items() if len(value) == 1]

        # Start with all nodes, gradually trim the leaves
        remaining_nodes = n
        print(edgesdict)  # Debug: print the adjacency list to verify structure
        
        # Perform a topological-like trim of the leaves
        while remaining_nodes > 2:
            # Decrease the count of remaining nodes by the number of leaves
            remaining_nodes -= len(leaves)
            new_leaves = []  # List to store new leaves after removing current ones

            # Remove current leaves from the graph
            for leaf in leaves:
                # Each leaf has only one neighbor, pop it
                neighbor = edgesdict[leaf].pop()
                # Remove the leaf from its neighbor's list
                edgesdict[neighbor].remove(leaf)

                # If the neighbor becomes a leaf (only one connection left), add it to new_leaves
                if len(edgesdict[neighbor]) == 1:
                    new_leaves.append(neighbor)

            # Update the list of leaves to the newly identified leaves
            leaves = new_leaves

        # Remaining nodes are the roots of the minimum height trees
        return leaves

File: 310-minimum-height-trees/minimum-height-trees.java
class Solution {

    public List<Integer> findMinHeightTrees(int n, int[][] edges) {
        if (n == 1) return new ArrayList<>(List.of(0));

        List<Set<Integer>> adjacentList = new ArrayList<>();

        for (int idx = 0; idx < n; idx++) adjacentList.add(new HashSet<>());

        for (int[] edge: edges){
            adjacentList.get(edge[0]).add(edge[1]);
            adjacentList.get(edge[1]).add(edge[0]);
        }

        List<Integer> leaves = new ArrayList<>();

        for (int idx = 0; idx < n; idx++) {
            if (adjacentList.get(idx).size() == 1) leaves.add(idx);
        }

        while (n > 2){
            n -= leaves.size();
            List<Integer> newLeaves = new ArrayList<>();

            for (int leaf: leaves){
                int nextLeaf = adjacentList.get(leaf).iterator().next();
                adjacentList.get(nextLeaf).remove(leaf);
                if (adjacentList.get(nextLeaf).size() == 1) newLeaves.add(nextLeaf);
            }

            leaves = newLeaves;
        }


        return leaves;


    }

}